secure/storage/src/namespaced_storage.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, GetResponse, KVStorage, Policy, Value};
+
+/// This provides a light wrapper around KV storages to support a namespace. That namespace is
+/// effectively prefixing all keys with then namespace value and "/" so a namespace of foo and a
+/// key of bar becomes "foo/bar". Without a namespace, the key would just be "bar". This matches
+/// how this library implements namespaces for Vault.
+pub struct NamespacedStorage<T> {
+    namespace: Option<String>,
+    inner: T,
+}
+
+impl<T: KVStorage> KVStorage for NamespacedStorage<T> {
+    fn available(&self) -> bool {
+        self.inner.available()
+    }
+
+    fn create(&mut self, key: &str, value: Value, policy: &Policy) -> Result<(), Error> {
+        self.inner.create(&self.ns_name(key), value, policy)
+    }
+
+    fn get(&self, key: &str) -> Result<GetResponse, Error> {
+        self.inner.get(&self.ns_name(key))
+    }
+
+    fn set(&mut self, key: &str, value: Value) -> Result<(), Error> {
+        self.inner.set(&self.ns_name(key), value)
+    }
+
+    /// Note: This is not a namespace function
```

Comment:
Doesn't this contradict what Vault already provides: https://github.com/libra/libra/blob/master/secure/storage/src/vault.rs#L44?

Reply:
These are not compatible, we do not support a list or delete operation at a KVStorage level which would be required to implement this generically.

I am okay with removing the specialized Vault namespace clean up and making it universal so that we could potentially leverage a common wrapper here.

Note: this is primarily intended for on disk and a potential "shared in-memory" storage solution at this point in time.

Motivation for not doing it in this PR? Simplicity, I don't really think that this functionality is critical or required for the work that I need out of this now, so minimizing bike shedding / extra work.

Reply:
Thanks, David! Your response makes sense to me. Ideally, I'd prefer for us to add this wrapper and integrate it with the Vault namespace implementation at the same time (i.e., in this in PR). But I see what you're saying about extra work.. My concern here is that we don't want someone doing something like "NamespacedStorage<VaultStorage>", because we'd have two namespace implementations... However, it might just be good enough to add a TODO  to address this if you want to move fast! I'll leave it up to you!